### DISCUSSION TITLE
any hints for debugging a bad diagram  (I assume that is the problem)

### DISCUSSION THREAD

**Discussion by macaroni57** [csgoh/processpiper#29] [Q&A] [answered]
I am generating a diagram from an excel file and getting an error in the .draw() function (see below).

Are there any hints for debugging a diagram?

Also, I think it would be handy to have a method on the process map object such that it generated properly indented code for drawing the diagram that was specified. That might allow me to determine where the error might lie.

Also, is there a requirement that there can only be one flow through the process nodes? Or can a set of process nodes belong to one flow and another set belongs to another flow? 

┌───────────────────── Traceback (most recent call last) ─────────────────────┐
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\spyder_kernels\py3comp │
│ at.py:356 in compat_exec                                                    │
│                                                                             │
│   353                                                                       │
│   354 def compat_exec(code, globals, locals):                               │
│   355 │   # Wrap exec in a function                                         │
│ > 356 │   exec(code, globals, locals)                                       │
│   357                                                                       │
│   358                                                                       │
│   359 if __name__ == '__main__':                                            │
│                                                                             │
│ c:\users\a205820\onedrive - fidelity technology group, llc\python           │
│ scripts\process_map.py:131 in <module>                                      │
│                                                                             │
│   128 │   │   │   │   │   node_count += 1                                   │
│   129 │   │   │   │   node_func[l].connect(node_func[n], '')                │
│   130                                                                       │
│ > 131 my_process_map.draw()                                                 │
│   132 my_process_map.save("Product_Feature_Interface_Process_Map.png")      │
│   133                                                                       │
│                                                                             │
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\processpiper\processma │
│ p.py:433 in draw                                                            │
│                                                                             │
│   430 │   │   │   │   │   │   f"Drawing connection for ({pool.name=}, {lane │
│   431 │   │   │   │   │   │   show_level="draw_connection",                 │
│   432 │   │   │   │   │   )                                                 │
│ > 433 │   │   │   │   │   lane.draw_connection(all_shapes)                  │
│   434 │   │                                                                 │
│   435 │   │   if self._footer is not None:                                  │
│   436 │   │   │   self._footer.draw()                                       │
│                                                                             │
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\processpiper\lane.py:2 │
│ 30 in draw_connection                                                       │
│                                                                             │
│   227 │   │   )                                                             │
│   228 │   │   if self.shapes:                                               │
│   229 │   │   │   shape = self.shapes[0]                                    │
│ > 230 │   │   │   shape.draw_connection(self.painter, all_shapes)           │
│   231 │                                                                     │
│   232 │   def set_draw_position(self, x: int, y: int, layout_grid: Grid) -> │
│   233 │   │   """Set the draw position of the lane"""                       │
│                                                                             │
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\processpiper\shape.py: │
│ 756 in draw_connection                                                      │
│                                                                             │
│   753 │   │   │   │   │   painter.connector_arrow_size,                     │
│   754 │   │   │   │   )                                                     │
│   755 │   │   │   │   next_shape = connection.target                        │
│ > 756 │   │   │   │   next_shape.draw_connection(painter, all_shapes)       │
│   757                                                                       │
│   758                                                                       │
│   759 @dataclass                                                            │
│                                                                             │
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\processpiper\shape.py: │
│ 756 in draw_connection                                                      │
│                                                                             │
│   753 │   │   │   │   │   painter.connector_arrow_size,                     │
│   754 │   │   │   │   )                                                     │
│   755 │   │   │   │   next_shape = connection.target                        │
│ > 756 │   │   │   │   next_shape.draw_connection(painter, all_shapes)       │
│   757                                                                       │
│   758                                                                       │
│   759 @dataclass                                                            │
│                                                                             │
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\processpiper\shape.py: │
│ 756 in draw_connection                                                      │
│                                                                             │
│   753 │   │   │   │   │   painter.connector_arrow_size,                     │
│   754 │   │   │   │   )                                                     │
│   755 │   │   │   │   next_shape = connection.target                        │
│ > 756 │   │   │   │   next_shape.draw_connection(painter, all_shapes)       │
│   757                                                                       │
│   758                                                                       │
│   759 @dataclass                                                            │
│                                                                             │
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\processpiper\shape.py: │
│ 756 in draw_connection                                                      │
│                                                                             │
│   753 │   │   │   │   │   painter.connector_arrow_size,                     │
│   754 │   │   │   │   )                                                     │
│   755 │   │   │   │   next_shape = connection.target                        │
│ > 756 │   │   │   │   next_shape.draw_connection(painter, all_shapes)       │
│   757                                                                       │
│   758                                                                       │
│   759 @dataclass                                                            │
│                                                                             │
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\processpiper\shape.py: │
│ 731 in draw_connection                                                      │
│                                                                             │
│   728 │   │   │   │   │   │   point_face_from,                              │
│   729 │   │   │   │   │   │   point_to,                                     │
│   730 │   │   │   │   │   │   point_face_to,                                │
│ > 731 │   │   │   │   │   ) = self.find_nearest_points_diff_pool(           │
│   732 │   │   │   │   │   │   connection.target, direction, all_shapes      │
│   733 │   │   │   │   │   )                                                 │
│   734 │   │   │   │   │   connection_style = "dashed"                       │
│                                                                             │
│ C:\Users\a205820\tools\Anaconda311\Lib\site-packages\processpiper\shape.py: │
│ 598 in find_nearest_points_diff_pool                                        │
│                                                                             │
│   595 │   │   │   │   │   }                                                 │
│   596 │   │                                                                 │
│   597 │   │   ### remove points from source and target shapes once they are │
│ > 598 │   │   del points_source[nearest_points["source_name"]]              │
│   599 │   │   del points_target[nearest_points["target_name"]]              │
│   600 │   │                                                                 │
│   601 │   │   return (                                                      │
└─────────────────────────────────────────────────────────────────────────────┘
KeyError: 'source_name'

**Comment by csgoh** [comment]
If you could share you code, I could help see where the problem is.

**Reply by macaroni57** [reply]
I believe this is identical to the dynamically generated code I was producing. This was a proof of concept so I have added other details.

from processpiper import ProcessMap, EventType, ActivityType, GatewayType

my_process_map = ProcessMap("Product Feature Interface", colour_theme="BLUEMOUNTAIN")
lane_system = my_process_map.add_lane('system')
node_SI = lane_system.add_element('SI', ActivityType.TASK)
node_Feature = lane_system.add_element('Feature', ActivityType.TASK)
node_Initiative = lane_system.add_element('Initiative', ActivityType.TASK)
node_Figma_Design = lane_system.add_element('Figma Design', ActivityType.TASK)
node_Story = lane_system.add_element('Story', ActivityType.TASK)
node_Product = lane_system.add_element('Product', ActivityType.TASK)
node_Bug = lane_system.add_element('Bug', ActivityType.TASK)
pool_Product_What = my_process_map.add_pool('Product What')
lane_Product_What_PrSL = pool_Product_What.add_lane('PrSL')
node_Product_feature_is_d = lane_Product_What_PrSL.add_element('Product feature is defined\nwith acceptance Criteria', ActivityType.TASK)
node_Product_feature_is_d.connect(node_SI, '')
node_Initiative.connect(node_Product_feature_is_d, '')
lane_Product_What_CSB = pool_Product_What.add_lane('CSB')
node_Product_feature_is_c = lane_Product_What_CSB.add_element('Product feature is claried in\nbusiness requirements', ActivityType.TASK)
node_Product_feature_is_c.connect(node_Feature, '')
node_SI.connect(node_Product_feature_is_c, '')
pool_Product_How = my_process_map.add_pool('Product How')
lane_Product_How_UXD__CSB = pool_Product_How.add_lane('UXD ; CSB')
node_Collaboratively_deve = lane_Product_How_UXD__CSB.add_element('Collaboratively develop UX\nDesigns with CSB', ActivityType.TASK)
node_Collaboratively_deve.connect(node_Figma_Design, '')
node_Feature.connect(node_Collaboratively_deve, '')
lane_Product_How_CSB__LRC = pool_Product_How.add_lane('CSB ; LRC')
node_Review_Figma_Design_ = lane_Product_How_CSB__LRC.add_element('Review Figma Design with LRC', ActivityType.TASK)
node_Review_Figma_Design_.connect(node_Figma_Design, '')
node_Figma_Design.connect(node_Review_Figma_Design_, '')
lane_Product_How_UXD = pool_Product_How.add_lane('UXD')
node_Update_Figma_Design_ = lane_Product_How_UXD.add_element('Update Figma Design based on\nLRC feedback', ActivityType.TASK)
node_Update_Figma_Design_.connect(node_Figma_Design, '')
node_Figma_Design.connect(node_Update_Figma_Design_, '')
lane_Product_How_Brokerage_Squad_UXD = pool_Product_How.add_lane('Brokerage Squad; UXD')
node_Work_with_Domains_SL = lane_Product_How_Brokerage_Squad_UXD.add_element('Work with Domains SL (Bob\nTaiani, Chris Coale) to\ndevelop features UI stories', ActivityType.TASK)
node_Work_with_Domains_SL.connect(node_Story, '')
node_Figma_Design.connect(node_Work_with_Domains_SL, '')
lane_Product_How_CSB = pool_Product_How.add_lane('CSB')
node_Create_feature_test_ = lane_Product_How_CSB.add_element('Create feature test cases', ActivityType.TASK)
node_Create_feature_test_.connect(node_Feature, '')
node_Feature.connect(node_Create_feature_test_, '')
lane_Product_How_Brokerage_Squad = pool_Product_How.add_lane('Brokerage Squad')
node_Implement_feature_UI = lane_Product_How_Brokerage_Squad.add_element('Implement feature UI stories', ActivityType.TASK)
node_Implement_feature_UI.connect(node_Product, '')
node_Story.connect(node_Implement_feature_UI, '')
node_Figma_Design.connect(node_Implement_feature_UI, '')
node_Verify_UI_in_virtual = lane_Product_How_UXD.add_element('Verify UI in virtual ui test\nenvironment implemented as\nspeced in Figma Design', ActivityType.TASK)
node_Verify_UI_in_virtual.connect(node_Bug, '')
node_Figma_Design.connect(node_Verify_UI_in_virtual, '')
node_Product.connect(node_Verify_UI_in_virtual, '')
node_Modify_screens_based = lane_Product_How_Brokerage_Squad.add_element('Modify screens based on UXD\nfeedback', ActivityType.TASK)
node_Modify_screens_based.connect(node_Product, '')
node_Bug.connect(node_Modify_screens_based, '')
node_Product.connect(node_Modify_screens_based, '')
node_Validate_that_Produc = lane_Product_How_CSB.add_element('Validate that Product meets\ncustomer needs per spec', ActivityType.TASK)
node_Validate_that_Produc.connect(node_Bug, '')
node_Feature.connect(node_Validate_that_Produc, '')
node_Modify_screens_based = lane_Product_How_Brokerage_Squad.add_element('Modify screens based on CSB\nfeedback', ActivityType.TASK)
node_Modify_screens_based.connect(node_Product, '')
node_Bug.connect(node_Modify_screens_based, '')


my_process_map.draw()
my_process_map.save("Product_Feature_Interface_Process_Map.png")

**Reply by csgoh** [reply, marked as answer]
Hi, there are two common issues related to this. 

1. One element can only have a maximum of 4 connections (incoming and outgoing). Some of the elements you added has more than 4. To overcome this, you can use gateway to join/merge multiple connections. Example: node_Feature, etc.
2. Element cannot be started on its own (i.e. start and end in the same element). Check node_Create_feature_test_ you have in your code)

I managed to get your code to work but not very pretty. The ProcessPiper is still in early days of development and some connections are overlapped and not drawn properly. I will try to fix it in the subsequence releases and update the document to reflect the maximum connection limit.

Here is the modified code:
```python
from processpiper import ProcessMap, EventType, ActivityType, GatewayType

with ProcessMap(
    "Product Feature Interface", colour_theme="BLUEMOUNTAIN"
) as my_process_map:
    with my_process_map.add_lane("system") as lane_system:
        node_Initiative = lane_system.add_element("Initiative", ActivityType.TASK)
        node_SI = lane_system.add_element("SI", ActivityType.TASK)
        node_Feature = lane_system.add_element("Feature", ActivityType.TASK)
        
        node_Figma_Design = lane_system.add_element("Figma Design", ActivityType.TASK)
        node_Story = lane_system.add_element('Story', ActivityType.TASK)
        node_Product = lane_system.add_element('Product', ActivityType.TASK)
        node_Bug = lane_system.add_element('Bug', ActivityType.TASK)

        g2 = lane_system.add_element("g2", GatewayType.INCLUSIVE)
        
        g2.connect(node_Figma_Design)
        g1 = lane_system.add_element("g1", GatewayType.INCLUSIVE)
        g1.connect(g2)
        
        g8 = lane_system.add_element("g8", GatewayType.INCLUSIVE)
        node_Feature.connect(g8, '')
        
        g9 = lane_system.add_element("g9", GatewayType.INCLUSIVE)
        g9.connect(node_Product, '')

        

    with my_process_map.add_pool("Product What") as pool_Product_What:
        with pool_Product_What.add_lane("PrSL") as lane_Product_What_PrSL:
            node_Product_feature_is_d = lane_Product_What_PrSL.add_element(
                "Product feature is defined\nwith acceptance Criteria",
                ActivityType.TASK,
            )
            node_Product_feature_is_d.connect(node_SI, "")
            node_Initiative.connect(node_Product_feature_is_d, "")

    with pool_Product_What.add_lane("CSB") as lane_Product_What_CSB:
        node_Product_feature_is_c = lane_Product_What_CSB.add_element(
            "Product feature is claried in\nbusiness requirements", ActivityType.TASK
        )
        g7 = lane_Product_What_CSB.add_element("g7", GatewayType.INCLUSIVE)
        node_Product_feature_is_c.connect(g7, "")
        g7.connect(node_Feature, '')
        node_SI.connect(node_Product_feature_is_c, "")

    with my_process_map.add_pool("Product How") as pool_Product_How:
        with pool_Product_How.add_lane("UXD ; CSB") as lane_Product_How_UXD__CSB:
            node_Collaboratively_deve = lane_Product_How_UXD__CSB.add_element(
                "Collaboratively develop UX\nDesigns with CSB", ActivityType.TASK
            )
            
            node_Collaboratively_deve.connect(g1, "")
            g8.connect(node_Collaboratively_deve, '')

        with pool_Product_How.add_lane("CSB ; LRC") as lane_Product_How_CSB__LRC:
            node_Review_Figma_Design_ = lane_Product_How_CSB__LRC.add_element(
                "Review Figma Design with LRC", ActivityType.TASK
            )
            node_Review_Figma_Design_.connect(g1, "")
            node_Figma_Design.connect(node_Review_Figma_Design_, "")

        with pool_Product_How.add_lane("UXD") as lane_Product_How_UXD:
            node_Update_Figma_Design_ = lane_Product_How_UXD.add_element(
                "Update Figma Design based on\nLRC feedback", ActivityType.TASK
            )
            node_Update_Figma_Design_.connect(g2, "")
            g3 = lane_Product_How_UXD.add_element("g3", GatewayType.INCLUSIVE)
            g3.connect(node_Update_Figma_Design_, "")
            

        with pool_Product_How.add_lane('Brokerage Squad; UXD') as lane_Product_How_Brokerage_Squad__UXD:
            node_Work_with_Domains_SL = lane_Product_How_Brokerage_Squad__UXD.add_element('Work with Domains SL (Bob\nTaiani, Chris Coale) to\ndevelop features UI stories', ActivityType.TASK)
            node_Work_with_Domains_SL.connect(node_Story, '')
            
            g3.connect(node_Work_with_Domains_SL)
            
            node_Figma_Design.connect(g3, '')
            

        with pool_Product_How.add_lane('CSB') as lane_Product_How_CSB:
            node_Create_feature_test_ = lane_Product_How_CSB.add_element('Create feature test cases', ActivityType.TASK)
            node_Create_feature_test_.connect(g7, '')
            g4 = lane_Product_How_CSB.add_element("g4", GatewayType.INCLUSIVE)
            g8.connect(g4, '')
            g4.connect(node_Create_feature_test_, '')

        with pool_Product_How.add_lane('Brokerage Squad') as lane_Product_How_Brokerage_Squad:
            node_Implement_feature_UI = lane_Product_How_Brokerage_Squad.add_element('Implement feature UI stories', ActivityType.TASK)
        
            node_Implement_feature_UI.connect(g9, '')
            node_Story.connect(node_Implement_feature_UI, '')
            
            g5 = lane_Product_How_Brokerage_Squad.add_element("g5", GatewayType.INCLUSIVE)
            g2.connect(g5)
            g5.connect(node_Implement_feature_UI, '')
        
            node_Verify_UI_in_virtual = lane_Product_How_UXD.add_element('Verify UI in virtual ui test\nenvironment implemented as\nspeced in Figma Design', ActivityType.TASK)
            g5.connect(node_Verify_UI_in_virtual, '')
            node_Verify_UI_in_virtual.connect(node_Bug, '')
            node_Product.connect(node_Verify_UI_in_virtual, '')

            node_Modify_screens_based = lane_Product_How_Brokerage_Squad.add_element('Modify screens based on UXD\nfeedback', ActivityType.TASK)
            node_Modify_screens_based.connect(g9, '')
            g6 = lane_Product_How_Brokerage_Squad.add_element("g6", GatewayType.INCLUSIVE)
            g6.connect(node_Modify_screens_based, '')
            node_Bug.connect(g6, '')
            node_Product.connect(g6, '')
            node_Validate_that_Produc = lane_Product_How_CSB.add_element('Validate that Product meets\ncustomer needs per spec', ActivityType.TASK)
            node_Validate_that_Produc.connect(node_Bug, '')
            node_Feature.connect(node_Validate_that_Produc, '')
            node_Modify_screens_based = lane_Product_How_Brokerage_Squad.add_element('Modify screens based on CSB\nfeedback', ActivityType.TASK)
            node_Modify_screens_based.connect(node_Product, '')
            node_Bug.connect(node_Modify_screens_based, '')

    my_process_map.draw()
    my_process_map.save("Product_Feature_Interface_Process_Map.png")

```

![image](https://github.com/csgoh/processpiper/assets/20867152/2c5fc84e-c5e7-4776-9a79-36b9152383ef)

**Reply by macaroni57** [reply]
Thank you so much for spending the time sorting through where I went astray. I figured there was some kind of constraint on the diagram but could not find anything in any docs I read.

I am not surprised by the circular references causing a problem. 

If I remove the "system" row, that greatly simplifies things which could be an option.

Best of luck as you continue to developing the product. It has real potential and I found it pretty easy to transform my spreadsheet representation into a diagram.
